numalogic/registry/redis_registry.py

```
@@ -0,0 +1,247 @@
+                    "model_version": model_version.decode(),
+                },
+            )
+        except ModelKeyNotFound as model_key_error:
+            _LOGGER.exception("Missing Key: %s", model_key_error)
+            return None
+        except RedisError as ex:
+            _LOGGER.exception("Unexpected error: %s", ex)
+            return None
+
+    def save(
+        self,
+        skeys: Sequence[str],
+        dkeys: Sequence[str],
+        artifact: artifact_t,
+        **metadata: str,
+    ) -> Optional[str]:
+        """
+        Saves the artifact into redis registry and updates version.
+        Args:
+            skeys: static key fields as list/tuple of strings
+            dkeys: dynamic key fields as list/tuple of strings
+            artifact: primary artifact to be saved
+            metadata: additional metadata surrounding the artifact that needs to be saved
+
+        Returns:
+            model version
+        """
+        try:
+            production_key = self.construct_key(skeys, dkeys, ["PRODUCTION"])
+            version = 0
+            if self.client.exists(production_key):
+                _LOGGER.info("Production key exists for the model")
+                version_key = self.client.get(name=production_key)
+                version = int(self.get_version(version_key.decode())) + 1
+            with self.client.pipeline() as pipe:
+                new_version_key = self.__save_artifact(pipe, artifact, skeys, dkeys, str(version))
+                if metadata:
+                    self.__save_metadata(pipe, metadata, new_version_key)
+                pipe.expire(name=new_version_key, time=self.ttl)
+                _LOGGER.info("Model is successfully with the key = %s", new_version_key)
+                pipe.execute()
+                return str(version)
+        except RedisError as ex:
+            _LOGGER.exception("Unexpected error: %s", ex)
+            return None
```

I don't think we should catch all exceptions here. Let's raise a custom exception

---

RedisError will only catch Redis exceptions.

Do we want to be more precise here?

---

This is the redis exceptions: https://github.com/redis/redis-py/blob/master/redis/exceptions.py

There can happen all the other kind of error and hence catching that. Open to ideas here.